substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+			}
+		}
+
+		log!(trace, "try-state: calculated approvals map: {:?}", approvals_map);
+
+		// compare calculated approvals per target with target list state.
+		for (target, calculated_stake) in approvals_map.iter() {
+			let stake_in_list = T::TargetList::get_score(target).expect("target must exist; qed.");
+			let stake_in_list = Self::to_vote_extended(stake_in_list);
+
+			if *calculated_stake != stake_in_list {
+				log!(
+						error,
+						"try-runtime: score of {:?} in `TargetList` list: {:?}, calculated sum of all stake: {:?}",
+						target,
+						stake_in_list,
+						calculated_stake,
+					);
+
+				return Err("target score in the target list is different than the expected".into())
+			}
+		}
+
+		frame_support::ensure!(
+			approvals_map.keys().count() == T::TargetList::iter().count(),
+			"calculated approvals count is different from total of target list.",
+		);
+
+		Ok(())
+	}
+
+	/// Try-state: checks if targets in the target list are sorted by score.
+	///
+	/// Invariant
+	///  * All targets in the target list are sorted by their score (approvals).
+	///
+	///  NOTE: unfortunatelly, it is not trivial to check if the sort correctness of the list if
+	///  the `SortedListProvider` is implemented by bags list due to score bucketing. Thus, we
+	///  leverage the [`SortedListProvider::in_position`] to verify if the target is in the
+	/// correct  position in the list (bag or otherwise), given its score.
+	pub fn do_try_state_target_sorting() -> Result<(), sp_runtime::TryRuntimeError> {
+		for t in T::TargetList::iter() {
+			frame_support::ensure!(
+				T::TargetList::in_position(&t).expect("target exists"),
+				"target list is not sorted"
+			);
+		}
+
+		Ok(())
+	}
+
+	/// Try-state: checks if voters in the voter list are sorted by score (stake).
+	///
+	/// Invariant
+	///  * All voters in the voter list are sorted by their score.
+	///
+	///  NOTE: unfortunatelly, it is not trivial to check if the sort correctness of the list if
+	///  the `SortedListProvider` is implemented by bags list due to score bucketing. Thus, we
+	///  leverage the [`SortedListProvider::in_position`] to verify if the target is in the
+	/// correct  position in the list (bag or otherwise), given its score.
+	pub fn do_try_state_voter_sorting() -> Result<(), sp_runtime::TryRuntimeError> {
+		for t in T::VoterList::iter() {
+			frame_support::ensure!(
+				T::VoterList::in_position(&t).expect("voter exists"),
+				"voter list is not sorted"
+			);
+		}
+
+		Ok(())
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// When a nominator's stake is updated, all the nominated targets must be updated
+	/// accordingly.
+	///
+	/// Note: it is assumed that `who`'s staking ledger state is updated *before* this method is
+	/// called.
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		match T::Staking::stake(who).and(T::Staking::status(who)) {
```

Comment:
Is `on_stake_update` ever fired for an account who will then not have `Staking::Stake` or `Stkaing::status`? I think this should be always implied, and therefore this is all defensive. 

Also, double checking if `T::Staking::stake(who)` is bringing any value here?

Reply:
> Is on_stake_update ever fired for an account who will then not have Staking::Stake or Stkaing::status? I think this should be always implied, and therefore this is all defensive.

Yes, that's correct. We assume that `on_stake_update` is only called on existing stashes so if staking interface returns an `Err` we move on and do defensive.

> Also, double checking if T::Staking::stake(who) is bringing any value here?
Not it doesn't really, removed it.